<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">

<head>
<title>README.html</title>
<meta http-equiv="Content-Type" content="application/xhtml+xml;charset=utf-8"/>
<link rel="stylesheet" type="text/css" media="all" href="https://cdn.jsdelivr.net/npm/github-markdown-css/github-markdown.min.css"  />
<link rel="stylesheet" type="text/css" media="all" href="https://cdn.jsdelivr.net/gh/highlightjs/cdn-release/build/styles/github.min.css"  /><meta name='viewport' content='width=device-width, initial-scale=1, shrink-to-fit=no'><style> body { box-sizing: border-box; max-width: 740px; width: 100%; margin: 40px auto; padding: 0 10px; } </style><script id='MathJax-script' async src='https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js'></script><script src='https://cdn.jsdelivr.net/gh/highlightjs/cdn-release/build/highlight.min.js'></script><script>document.addEventListener('DOMContentLoaded', () => { document.body.classList.add('markdown-body'); document.querySelectorAll('pre[lang] > code').forEach((code) => { code.classList.add(code.parentElement.lang); }); document.querySelectorAll('pre > code').forEach((code) => { hljs.highlightBlock(code); }); });</script>
</head>

<body>

<h1 id="haplopy-haplotype-estimation-and-phasing-in-python">HaploPy – Haplotype estimation and phasing in Python</h1>
<p>This package contains tools for estimating haplotype (or allele list) frequencies in a population using measurements of unphased genotype data, that is, phenotypes.</p>
<h2 id="introduction">Introduction</h2>
<p>In layman terms, a phenotype is defined as an observation of two-allele sets over multiple gene loci:</p>
<pre class="shell"><code>    Aa ––––––– Bb ––––––– Cc
    |          |          |
  locus 1    locus 2    locus 3</code></pre>
<p>Note that the above datum doesn’t reveal what are the exact haplotype (allele sequence) pair behind the phenotype. Possible parent haplotype pairs that could result into the above phenotype are given by</p>
<pre class="shell"><code>(ABC, abc), (aBC, Abc), (AbC, aBc), (abC, ABc)</code></pre>
<p>In other words, the mapping that maps a haplotype pair to a phenotype is defined by the example</p>
<pre class="shell"><code>(Abc, aBC) =&gt; (Aa, Bb, Cc)</code></pre>
<p>and so on. Note that each item in the phenotype is a set of two alleles where the order doesn’t matter.</p>
<p><strong>Problem:</strong> Suppose that we have a set of phenotype observations from a large population of N individuals. For each individual phenotype we would like to estimate what are the most probable haplotype pair that resulted into the phenotype. The main ingredient of solution is the estimation of individual haplotype frequencies in the population</p>
<h2 id="installation">Installation</h2>
<p>The package is found in PyPi.</p>
<pre class="shell"><code>pip install haplopy</code></pre>
<h2 id="examples">Examples</h2>
<h3 id="estimate-haplotype-frequencies">Estimate haplotype frequencies</h3>
<p>Simulate dataset using a prescribed haplotype probabilities and a multinomial distribution model.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1"></a><span class="im">import</span> haplopy <span class="im">as</span> hp</span>
<span id="cb5-2"><a href="#cb5-2"></a></span>
<span id="cb5-3"><a href="#cb5-3"></a></span>
<span id="cb5-4"><a href="#cb5-4"></a>p_haplotypes <span class="op">=</span> {</span>
<span id="cb5-5"><a href="#cb5-5"></a>    (<span class="st">&quot;A&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;C&quot;</span>): <span class="fl">0.34</span>,</span>
<span id="cb5-6"><a href="#cb5-6"></a>    (<span class="st">&quot;a&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;c&quot;</span>): <span class="fl">0.20</span>,</span>
<span id="cb5-7"><a href="#cb5-7"></a>    (<span class="st">&quot;a&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;C&quot;</span>): <span class="fl">0.13</span>,</span>
<span id="cb5-8"><a href="#cb5-8"></a>    (<span class="st">&quot;a&quot;</span>, <span class="st">&quot;b&quot;</span>, <span class="st">&quot;c&quot;</span>): <span class="fl">0.23</span>,</span>
<span id="cb5-9"><a href="#cb5-9"></a>    (<span class="st">&quot;A&quot;</span>, <span class="st">&quot;b&quot;</span>, <span class="st">&quot;C&quot;</span>): <span class="fl">0.10</span></span>
<span id="cb5-10"><a href="#cb5-10"></a>}</span>
<span id="cb5-11"><a href="#cb5-11"></a></span>
<span id="cb5-12"><a href="#cb5-12"></a>phenotypes <span class="op">=</span> hp.multinomial.Model(p_haplotypes).random(<span class="dv">100</span>)</span>
<span id="cb5-13"><a href="#cb5-13"></a></span>
<span id="cb5-14"><a href="#cb5-14"></a>fig <span class="op">=</span> hp.plot.plot_haplotypes(p_haplotypes)</span>
<span id="cb5-15"><a href="#cb5-15"></a>fig <span class="op">=</span> hp.plot.plot_phenotypes(phenotypes)</span></code></pre></div>
<figure>
<img src="./doc/images/hinton-original.png" title="Original" alt="" /><figcaption>Original relative haplotype frequencies</figcaption>
</figure>
<figure>
<img src="./doc/images/bar.png" title="Phenotypes" alt="" /><figcaption>Simulated phenotype observation set</figcaption>
</figure>
<p>Pretend that we don’t know the underlying haplotype distribution and let’s try to estimate it.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1"></a>model <span class="op">=</span> hp.multinomial.Model().fit(phenotypes)</span>
<span id="cb6-2"><a href="#cb6-2"></a>fig <span class="op">=</span> hp.plot.plot_haplotypes(</span>
<span id="cb6-3"><a href="#cb6-3"></a>    model.p_haplotypes,</span>
<span id="cb6-4"><a href="#cb6-4"></a>    thres<span class="op">=</span><span class="fl">1.0e-6</span>  <span class="co"># Hide probabilities smaller than this</span></span>
<span id="cb6-5"><a href="#cb6-5"></a>)</span></code></pre></div>
<figure>
<img src="./doc/images/hinton-estimated.png" title="Estimated" alt="" /><figcaption>Estimated relative haplotype frequencies</figcaption>
</figure>
<h2 id="to-be-added-features">To-be-added features</h2>
<ul>
<li><strong>TODO</strong> Phasing of unphased phenotypes given a haplotype frequency model</li>
<li><strong>TODO</strong> Imputation of missing locus measurements</li>
</ul>

</body>
</html>
